File: Problems/0010_Regular_Expression_Matching/Regular_Expression_Matching.py
import os
import sys
import time
from collections import deque

class Solution:
    def isMatch(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: bool
        """

def main():
    argv = sys.argv
    argc = len(argv)

    if (argc < 2):
        print("Usage: python %s <testdata.txt>" %(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("%s not found..." %argv[1])
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        print("argv[1] = %s" %temp)
        loop_main(temp)
    #    print("Hit Return to continue...")
    #    input()


def loop_main(temp):
    flds = temp.replace("\"","").replace("[","").replace("]","").rstrip().split(",")
    s = flds[0]
    p = flds[1]
    print("s = %s, p = %s" %(s, p)

    time0 = time.time()

    sl = Solution()
    result = sl.isMatch(s, p)

    print("result = %s" %result)

    time1 = time.time()
    print("Execute time ... : %f[s]" %(time1 - time0))
    print()


if __name__ == "__main__":
    main()

File: Problems/0010_Regular_Expression_Matching/Regular_Expression_Matching.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Solution
{
    public bool IsMatch(string s, string p)
    {
        return IsMatch(s, p, 0, 0);
    }
    
    bool IsMatch(string s, string p, int i, int j)
    {    
        //base case - reached end of pattern
        if (j >= p.Length)
        {
            return i >= s.Length && j >= p.Length; 
        }
        
        if (j + 1 < p.Length && p[j + 1] == '*')
        {   //peek ahead for *
            while (i < s.Length && (s[i] == p[j] || p[j] == '.'))
            { 
                if (IsMatch(s, p, i, j + 2))
                    return true;
                i++;
            }
            return IsMatch(s, p, i, j + 2);
        }
        else if ( i < s.Length && ( s[i] == p[j] || p[j] == '.' ) )
        {   //direct 1-to-1 match
            return IsMatch(s, p, i + 1, j + 1);
        }
        
        return false;
    }

    public void Main(string args)
    {
        string[] flds = args.Replace("\"","").Replace("[","").Replace("]","").Trim().Split(',');
        string s = flds[0];
        string p = flds[1];

        Console.WriteLine("s = " + s + ", p = " + p);

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        bool result = IsMatch(s, p);
        Console.WriteLine("result = " + result.ToString());
        
        sw.Stop();
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms");
    }
}
